### PR TITLE
Add TLS Registry configuration to WebSockets Next Client

### LINKED ISSUE #41004: Add support for the TLS registry to the WebSocket-Next client extension
### Description

With the integrated TLS registry, it should be possible to configure the WebSocket Next client using the TLS registry instead of the specific configuration.

### Implementation ideas

This is the code used for the mailer:

```java
 private void configureTLS(String name, MailerRuntimeConfig config, TlsConfigurationRegistry tlsRegistry, MailConfig cfg,
            boolean globalTrustAll) {
        TlsConfiguration configuration = null;

        // Check if we have a named TLS configuration or a default configuration:
        if (config.tlsConfigurationName.isPresent()) {
            Optional<TlsConfiguration> maybeConfiguration = tlsRegistry.get(config.tlsConfigurationName.get());
            if (!maybeConfiguration.isPresent()) {
                throw new IllegalStateException("Unable to find the TLS configuration "
                        + config.tlsConfigurationName.get() + " for the mailer " + name + ".");
            }
            configuration = maybeConfiguration.get();
        } else if (tlsRegistry.getDefault().isPresent() && tlsRegistry.getDefault().get().isTlsEnabled()) {
            configuration = tlsRegistry.getDefault().get();
        }

       // Apply the configuration
        if (configuration != null) {
            // This part is often the same (or close) for every Vert.x client:
            cfg.setSsl(true);

            if (configuration.getTrustStoreOptions() != null) {
                cfg.setTrustOptions(configuration.getTrustStoreOptions());
            }

           // For mTLS:
            if (configuration.getKeyStoreOptions() != null) {
                cfg.setKeyCertOptions(configuration.getKeyStoreOptions());
            }

            if (configuration.isTrustAll()) {
                cfg.setTrustAll(true);
            }
            if (configuration.getHostnameVerificationAlgorithm().isPresent()) {
               // ACHTUNG HERE - this is protocol specific. The HTTP-based protocols should use HTTPS by default. 
                cfg.setHostnameVerificationAlgorithm(configuration.getHostnameVerificationAlgorithm().get());
            }

            SSLOptions sslOptions = configuration.getSSLOptions();
            if (sslOptions != null) {
                cfg.setSslHandshakeTimeout(sslOptions.getSslHandshakeTimeout());
                cfg.setSslHandshakeTimeoutUnit(sslOptions.getSslHandshakeTimeoutUnit());
                for (String suite : sslOptions.getEnabledCipherSuites()) {
                    cfg.addEnabledCipherSuite(suite);
                }
                for (Buffer buffer : sslOptions.getCrlValues()) {
                    cfg.addCrlValue(buffer);
                }
                cfg.setEnabledSecureTransportProtocols(sslOptions.getEnabledSecureTransportProtocols());

            }

        } else {
           // Mailer specific configuration (very incomplete as you can see:
            boolean trustAll = config.trustAll.isPresent() ? config.trustAll.get() : globalTrustAll;
            cfg.setSsl(config.ssl);
            cfg.setTrustAll(trustAll);
            applyTruststore(config, cfg);
        }
    }
```

### PR DESCRIPTION
**Pull request by geoand**
- Resolves: #41004

~~P.S. Without support in the server part, I am not sure what the best way to test this is~~

### CODE REVIEW

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java [hunk 1]
```diff
@@ -63,6 +66,12 @@ public Uni<WebSocketClientConnection> connect() {
             clientOptions.setMaxMessageSize(config.maxMessageSize().getAsInt());
         }
 
+        Optional<TlsConfiguration> tlsConfiguration = TlsConfiguration.from(tlsConfigurationRegistry,
```
Comment (mkouba) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java:69]: Shouldn't we do the same in the `BasicWebSocketConnectorImpl#connect()`?

**Comment by mkouba** [comment]
> P.S. Without support in the server part, I am not sure what the best way to test this is

That's a pity. What needs to be done on the server?

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java (hunk 1 above)
Comment (geoand) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java:69]: Ah yeah, I forgot to add it there - will do tomorrow

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java [hunk 2]
```diff
@@ -63,6 +66,12 @@ public Uni<WebSocketClientConnection> connect() {
...
+                config.tlsConfigurationName());
+        if (tlsConfiguration.isPresent()) {
+            clientOptions.setTrustOptions(tlsConfiguration.get().getTrustStoreOptions());
```
Comment (cescoffier) [inline extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java:72]: There is a bit more to do:

- alpn, handshake timeout, trust-all...

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java (hunk 2 above)
Comment (geoand) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java:72]: I will add them tomorrow

**Comment by geoand** [comment]
> > P.S. Without support in the server part, I am not sure what the best way to test this is
> 
> That's a pity. What needs to be done on the server?

Support TLS on the server as well :)

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java (hunk 1 above)
Comment (geoand) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java]: Added

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java (hunk 2 above)
Comment (geoand) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorImpl.java]: Added

**Comment by cescoffier** [comment]
The server is using the primary HTTP server which has been refactored already a few weeks ago (when we merged the TLS registry)

**Comment by geoand** [comment]
> The server is using the primary HTTP server which has been refactored already a few weeks ago (when we merged the TLS registry)

Cool, I'll add some tests then

**Comment by cescoffier** [comment]
@geoand check how the tests have been made in Quarkus Vert.x HTTP. We generate (tons of) certificates automatically.

**Comment by geoand** [comment]
PR updated with tests

**Comment by cescoffier** [comment]
BTW, I wanted to add an important note here: For WebSocket, the default hostname verifier algorithm **must** be `HTTPS`. The user has the possibility to set it to `NONE`, but it must be done conscientiously.

**Comment by geoand** [comment]
Right, that's how things are currently handled, no?

**Review by cescoffier — APPROVED** [review]
Just one tiny comment.

File: extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorBase.java [hunk 3]
```diff
@@ -129,4 +139,58 @@ String replacePathParameters(String path) {
...
+        Optional<TlsConfiguration> maybeTlsConfiguration = TlsConfiguration.from(tlsConfigurationRegistry,
+                config.tlsConfigurationName());
+        if (maybeTlsConfiguration.isPresent()) {
+            clientOptions.setSsl(true);
+
+            TlsConfiguration tlsConfiguration = maybeTlsConfiguration.get();
+            clientOptions.setTrustOptions(tlsConfiguration.getTrustStoreOptions());
+
+            if (tlsConfiguration.getTrustStoreOptions() != null) {
+                clientOptions.setTrustOptions(tlsConfiguration.getTrustStoreOptions());
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketConnectorBase.java:163]: This was an oversight, I'll fix it

**Pull request merged**